microcosm/opaque.py

```diff
@@ -29,6 +29,39 @@
 )
 
 
+class LowercaseDict(dict):
+    """
+    Dict where all str keys are lowercase
+
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        for key in list(self.keys()):
+            value = super().pop(key)
+            self[key] = value
+
+    def __setitem__(self, key, value):
+        key = key.casefold() if isinstance(key, str) else key
+        super().__setitem__(key, value)
+
+    def update(self, *args, **kwargs):
+        for key, value in dict(*args, **kwargs).items():
+            self[key] = value
+
+    def setdefault(self, key, *args, **kwargs):
+        key = key.casefold() if isinstance(key, str) else key
+        super().setdefault(key, *args, **kwargs)
+
+    @classmethod
+    def fromkeys(cls, keys, v=None):
+        keys = (
+            key.casefold() if isinstance(key, str) else key
+            for key in keys
+        )
+        return super().fromkeys(keys, v)
+
```

Comment:
I'm wondering if we also want to make the __getitem__ also normalise the key input?

Reply:
Possibly, won't hurt, but can help with backwards compatibility. I will add it.